main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,240 @@
+import ca.uwaterloo.flix.language.ast.{OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.ParOps
+
+/**
+  * The occurrence analyzer collects occurrence information on binders according to the definition of [[Occur]].
+  * Additionally, it also counts the number of subexpressions in a function to compute its size.
+  */
+object OccurrenceAnalyzer {
+
+  /**
+    * Performs occurrence analysis on the given AST `root`.
+    */
+  def run(root: OccurrenceAst.Root)(implicit flix: Flix): OccurrenceAst.Root = {
+    val (ds, os) = ParOps.parMap(root.defs.values)(visitDef).unzip
+
+    // Combine all `defOccurrences` into one map.
+    val defOccur = combineAll(os)
+
+    // Updates the occurrence of every `def` in `ds` based on the occurrence found in `defOccur`.
+    val defs = ds.foldLeft(Map.empty[DefnSym, OccurrenceAst.Def]) {
+      case (macc, defn) =>
+        val occur = defOccur.getOrElse(defn.sym, Dead)
+        val newContext = defn.context.copy(occur = occur)
+        val defWithContext = defn.copy(context = newContext)
+        macc + (defn.sym -> defWithContext)
+    }
+    OccurrenceAst.Root(defs, root.enums, root.structs, root.effects, root.mainEntryPoint, root.entryPoints, root.sources)
+  }
+
+  /**
+    * Performs occurrence analysis on `defn`.
+    */
+  private def visitDef(defn0: OccurrenceAst.Def): (OccurrenceAst.Def, ExpContext) = {
+    val (exp, expCtx) = visitExp(defn0.exp)(defn0.sym)
+    val defContext = DefContext(expCtx.get(defn0.sym), expCtx.size, expCtx.localDefs, isDirectCall(exp), isSelfRecursive(expCtx, defn0))
+    val fparams = defn0.fparams.map(fp => fp.copy(occur = expCtx.get(fp.sym)))
+    (OccurrenceAst.Def(defn0.sym, fparams, defn0.spec, exp, defContext, defn0.loc), expCtx)
+  }
+
+  /**
+    * Returns true if `expr0` is a function call.
+    */
+  private def isDirectCall(expr0: OccurrenceAst.Expr): Boolean = expr0 match {
+    case OccurrenceAst.Expr.ApplyDef(_, _, _, _, _, _) => true
+    case OccurrenceAst.Expr.ApplyClo(_, _, _, _, _) => true
+    case _ => false
+  }
+
+  /**
+    * Returns true if `def0` occurs in `expCtx`.
+    */
+  private def isSelfRecursive(expCtx: ExpContext, defn0: OccurrenceAst.Def): Boolean = expCtx.defs.get(defn0.sym) match {
+    case None => false
+    case Some(o) => o match {
+      case Occur.Dead => false
+      case Occur.Once => true
+      case Occur.OnceInLambda => true
+      case Occur.OnceInLocalDef => true
+      case Occur.Many => true
+      case Occur.ManyBranch => true
+      case Occur.DontInline => true
+      case Occur.DontInlineAndDontRewrite => true
+    }
+  }
+
+  /**
+    * Performs occurrence analysis on `exp0`
+    */
+  private def visitExp(exp0: OccurrenceAst.Expr)(implicit sym0: Symbol.DefnSym): (OccurrenceAst.Expr, ExpContext) = (exp0, ExpContext.empty)
+
+  /**
+    * Combines `expCtx1` and `expCtx2` into a single [[ExpContext]].
+    */
+  private def combineInfoBranch(expCtx1: ExpContext, expCtx2: ExpContext): ExpContext = {
```

Comment:
There are two types right? Can we reflect that in names:

combineSeq
combine???